python/tests/tornado/test_hyperbahn.py

```diff
@@ -57,8 +58,16 @@ def test_request():
 
 
 @pytest.mark.gen_test
-def test_register():
-    channel = TChannel(name='test')
+def test_advertise(tchannel_server):
+    endpoint = b'ad'
 
-    with pytest.raises(ConnectionClosedError):
-        yield hyperbahn.advertise(channel, 'foo', ['127.0.0.1:23000'])
+    @tornado.gen.coroutine
+    def execution(request, response):
+        body = yield response.get_body()
+        assert body['services']['serviceName'] == 'test'
+        assert body['services']['cost'] == 0
+    tchannel_server.expect_call(endpoint, 'json').execution = execution
+    channel = TChannel(name='test')
+    hyperbahn.advertise(channel,
+                        'test', ['localhost:'+str(tchannel_server.port)]
+                        )
```

Comment:
The test is wrong.

- `advertise` returns a future. You're not waiting for the future to return. This function will exit immediately. You won't have tested anything.
- `Expectation.execution` isn't a thing. It's `Expectation.execute`. However you're not supposed to assign it like that anyway. That whole class is an artifact from before we had a convenient way to set up ad-hoc servers.
- `response.get_body`? Should be `request`.

Here's what I think you should be doing instead of trying to shoehorn this behavior into the `TestServer`.

    server = TChannel('test-server')

    @server.register('ad', 'json')
    def ad(req, res):
        # ...

    server.listen()
    client = TChannel('test-client')
    client.advertise(..., [server.hostport])

Also, asserting values inside the request handler is useless if you don't assert that the handler was actually called. If the handler is never called, your test will pass regardless. That's why I told you to use a mock object for that. You could make sure it was actually called:

    call = mock.Mock()

    @tornado.gen.coroutine
    def execute(req, res):
        body = # ...
        call(body)

     call.assert_called_once_with({...})